staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -552,3 +547,25 @@ func (r *inMemoryResponseWriter) String() string {
 	}
 	return s
 }
+
+// isV2Beta1ContentType checks of the content-type string is both
+// "application/json" and contains the v2beta1 content-type params.
+// NOTE: This function is resilient to the ordering of the
+// content-type parameters, as well as parameters added by
+// intermediaries such as proxies or gateways. Examples:
+//
+//	"application/json; g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList" = true
+//	"application/json; as=APIGroupDiscoveryList;v=v2beta1;g=apidiscovery.k8s.io" = true
+//	"application/json; as=APIGroupDiscoveryList;v=v2beta1;g=apidiscovery.k8s.io;charset=utf-8" = true
+//	"application/json" = false
+//	"application/json; charset=UTF-8" = false
+func isV2Beta1ContentType(contentType string) bool {
```

Comment:
Duplicated from client-go to avoid adding extra dependencies and make backport easier. Going forward, is there a better spot for this function?

Reply:
Elsewhere in this file we also have

```go
req.Header.Add("Accept", runtime.ContentTypeJSON+";g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList")
```

I think we should keep the logic for what we expect and what request together. So if this function is exported, then the requested header should also use a constant. But I don't like idea of exposing a public `IsV2Beta1ContentType` since new versions would need a lot of new plumbing, and it kinda leaks implementation details.

I think it would make sense to expose from client-go or maybe a shared util package a function like:

```go
 // ContentTypeIsGVK checks of the content-type string is both
 // "application/json" and matches the provided GVK.
 // NOTE: This function is resilient to the ordering of the
 // content-type parameters, as well as parameters added by
 // intermediaries such as proxies or gateways. Examples:
 //
 //	("application/json; g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList", {apidiscovery.k8s.io, v2beta1, APIGroupDiscoveryList}) = true
 //	("application/json; as=APIGroupDiscoveryList;v=v2beta1;g=apidiscovery.k8s.io", {apidiscovery.k8s.io, v2beta1, APIGroupDiscoveryList}) = true
 //	("application/json; as=APIGroupDiscoveryList;v=v2beta1;g=apidiscovery.k8s.io;charset=utf-8", {apidiscovery.k8s.io, v2beta1, APIGroupDiscoveryList}) = true
 //	("application/json", any GVK) = false
 //	("application/json; charset=UTF-8", any GVK) = false
 //
func ContentTypeIsGVK(string, schema.GroupVersionKind) bool
```

which performs this check for any content type. That way we wouldnt add to client-go api a function we are likely to remove, and we can keep our logic for what GVK is expected in the same place as our logic as what is requested: If the server switches to a new supported versions we only have one place to update.

Also considered something like `GVKFromContentType(string) (schema.GroupVersionKind, error)`. Which I find to be marginally less ergonomic due to the possible error return.

Reply:
> Duplicated from client-go to avoid adding extra dependencies and make backport easier.

client-go is already a dependency of kube-aggregator. Is another dependency triggered if you try to add it?

https://github.com/kubernetes/kubernetes/blob/8f383599a810729176c385e91d4f511861fc962f/staging/src/k8s.io/kube-aggregator/go.mod#L2-L19

Reply:
I like the ContentTypeIsGVK function, implemented. We would have to expose `IsV2Beta1ContentType` somewhere though? Or they can be instantiated individually (which is what is done at the moment), but that creates more room for the gvks to accidentally not match?

When we release a new version the client shouldn't drop support for v2beta1 and it would be something like `Accepts: application/json; gvk for v1, application/json gvk for v2beta1`